contracts/PrepaidCardMarket_v2.sol

```diff
@@ -0,0 +1,304 @@
+  );
+  event TrustedProvisionerAdded(address token);
+  event TrustedProvisionerRemoved(address token);
+  event PausedToggled(bool paused);
+
+  function initialize(address owner) public override initializer {
+    paused = false;
+    OwnableInitialize(owner);
+  }
+
+  function setup(
+    address _exchangeAddress,
+    address _prepaidCardManagerAddress,
+    address _provisioner,
+    address _tokenManager,
+    address[] calldata _trustedProvisioners,
+    address _versionManager
+  ) external onlyOwner {
+    exchangeAddress = _exchangeAddress;
+    prepaidCardManagerAddress = _prepaidCardManagerAddress;
+    provisioner = _provisioner;
+    tokenManager = _tokenManager;
+    versionManager = _versionManager;
+
+    for (uint256 i = 0; i < _trustedProvisioners.length; i++) {
+      _addTrustedProvisioner(_trustedProvisioners[i]);
+    }
+
+    emit Setup();
+  }
+
+  function setPaused(bool _paused) external onlyOwner {
+    paused = _paused;
+    emit PausedToggled(_paused);
+  }
+
+  function getTrustedProvisioners() external view returns (address[] memory) {
+    return trustedProvisioners.values();
+  }
+
+  function _addTrustedProvisioner(address _token) internal {
+    trustedProvisioners.add(_token);
+    emit TrustedProvisionerAdded(_token);
+  }
+
+  function removeTrustedProvisioner(address _token) external onlyOwner {
+    trustedProvisioners.remove(_token);
+    emit TrustedProvisionerRemoved(_token);
+  }
+
+  function provisionPrepaidCard(address customer, bytes32 sku)
+    external
+    returns (bool)
+  {
+    require(!paused, "Contract is paused");
+    require(
+      trustedProvisioners.contains(msg.sender),
+      "Only trusted provisioners allowed"
+    );
+    require(asks[sku] > 0, "Can't provision SKU with 0 askPrice");
+
+    PrepaidCardManager prepaidCardManager = PrepaidCardManager(
+      prepaidCardManagerAddress
+    );
+
+    address token = skus[sku].issuingToken;
+    uint256 faceValue = skus[sku].faceValue;
+    string memory customizationDID = skus[sku].customizationDID;
+    uint256 priceToCreatePrepaidCard = prepaidCardManager.priceForFaceValue(
+      token,
+      faceValue
+    ); // in Wei
+
+    uint256[] memory issuingTokenAmounts = new uint256[](1);
+    uint256[] memory spendAmounts = new uint256[](1);
+
+    issuingTokenAmounts[0] = Exchange(exchangeAddress).convertFromSpend(
+      token,
+      faceValue
+    );
+
+    spendAmounts[0] = faceValue;
+
+    address issuer = skus[sku].issuer;
+    address issuerSafe = skus[sku].issuerSafe;
+
+    IERC677(token).safeTransferAndCall(
+      prepaidCardManagerAddress,
+      priceToCreatePrepaidCard,
+      abi.encode(
+        customer,
+        issuingTokenAmounts,
+        spendAmounts,
+        customizationDID,
+        address(0), // marketAddress - we probably don't need it in this case?
+        issuer,
+        issuerSafe
+      )
+    );
+
+    balance[issuerSafe][token] -= priceToCreatePrepaidCard;
+
+    return true;
```

Comment:
I think we are missing a ProvisionPrepaidCard event, can you make sure that is part of the unit tests too?

Reply:
Do you have any suggestions on what would be the most proper way of emitting this in the `PrepaidCardMarketV2`? It doesn't look so straightforward to me because when we send the tokens using `IERC677(token).safeTransferAndCall` in `provisionPrepaidCard` method, we don't get the prepaid card address back, meaning we can not emit `ProvisionedPrepaidCard(prepaidCard, customer, sku, askPrice)` at this spot.

So I'm wondering where would be good to emit this event. One option is to do it in the `PrepaidCardManager` in the `onTokenTransfer` hook after creating the card, but for that we'd also need to send the `sku` and `askPrice` in the `data`. One other option I see is to provide a `onPrepaidCardCreate` callback function in the V2 contract which the manager would call. Is it maybe possible to somehow query for the newly created prepaid card address in the end of the `provisionPrepaidCard` function in the V2 contract?

Reply:
That’s fine to not have the prepaid card address. The prepaid card manager will emit an event with this address in the same txn that the subgraph can correlate . This can be an event that has different args. Specifically you can include the SKU and the prepaid card owner. It is important though that any publicly facing function always emits an event—it absolutely something that we are audited for.